packages/gatsby-plugin-mdx/src/compile-mdx.ts

```diff
@@ -1,27 +1,108 @@
+import type { NodePluginArgs } from "gatsby"
 import type { ProcessorOptions } from "@mdx-js/mdx"
-import type { IFileNode, IMdxNode } from "./types"
+import type { IFileNode, IMdxMetadata, IMdxNode } from "./types"
+
+import deepmerge from "deepmerge"
+
+import { enhanceMdxOptions, IMdxPluginOptions } from "./plugin-options"
 
 // Compiles MDX into JS
 // This could be replaced by @mdx-js/mdx if MDX compile would
 // accept custom data passed to the unified pipeline via processor.data()
-export default async function compileMDX(
-  source: string,
+export async function compileMDX(
   mdxNode: IMdxNode,
   fileNode: IFileNode,
   options: ProcessorOptions
-): Promise<string> {
-  const { createProcessor } = await import(`@mdx-js/mdx`)
-  const { VFile } = await import(`vfile`)
+): Promise<{ processedMDX: string; metadata: IMdxMetadata }> {
+  try {
+    const { createProcessor } = await import(`@mdx-js/mdx`)
+    const { VFile } = await import(`vfile`)
+
+    const processor = createProcessor(options)
+
+    // If we could pass this via MDX loader config, this whole custom loader might be obsolete.
+    const metadata: IMdxMetadata = {}
+    processor.data(`mdxNode`, mdxNode)
+    processor.data(`mdxMetadata`, metadata)
 
-  const processor = createProcessor(options)
+    const result = await processor.process(
+      // Inject path to original file for remark plugins. See: https://github.com/gatsbyjs/gatsby/issues/26914
+      new VFile({ value: mdxNode.body, path: fileNode.absolutePath })
+    )
 
-  // If we could pass this via MDX loader config, this whole custom loader might be obsolete.
-  processor.data(`mdxNode`, mdxNode)
+    // Clone metadata so ensure it won't be overridden by later processings
+    const clonedMetadata = Object.assign(
+      {},
+      processor.data(`mdxMetadata`) as IMdxMetadata
+    )
+    const processedMDX = result.toString()
 
-  const result = await processor.process(
-    // Inject path to original file for remark plugins. See: https://github.com/gatsbyjs/gatsby/issues/26914
-    new VFile({ value: source, path: fileNode.absolutePath })
+    return { processedMDX, metadata: clonedMetadata }
+  } catch (err) {
+    const errorMeta = [
```

Comment:
This feels a bit verbose. Wouldn't the absolute path to the file suffice to annotate the error?

Reply:
Also, we should use the `reporter` here instead of just `throw err`, e.g. use `reporter.panicOnBuild()` with the simplified message

Reply:
Done!